### PR TITLE
Switch to rust-tongsuo, supporting SM2 and SM4 algorithms.

### PR DESCRIPTION
**Pull request by wa5i**

### CODE REVIEW

File: src/utils/key.rs [hunk 1]
```diff
@@ -51,43 +70,56 @@ impl KeyBundle {
...
-                let pkey = PKey::from_ec_key(ec_key)?;
-                pkey.private_key_to_pem_pkcs8()?
-            }
-            "aes-gcm" | "aes-cbc" | "aes-ecb" => {
-                if key_bits != 128 && key_bits != 192 && key_bits != 256 {
-                    return Err(RvError::ErrPkiKeyBitsInvalid);
+                let ec_key = EcKey::generate(&ec_group)?;
+                PKey::from_ec_key(ec_key)?.private_key_to_pem_pkcs8()?
+            },
+            "sm2" => {
```
Comment (InfoHunter) [inline src/utils/key.rs:93]: It seems this needs `#[cfg(not(tongsuo))]` as well?

Comment (InfoHunter) [reply src/utils/key.rs:93]: Unless you may compile this code only with OpenSSL 1.1.1 or later. By luck that users are not using an older version of OpenSSL.

File: src/utils/key.rs [hunk 2]
```diff
@@ -99,95 +131,65 @@ impl KeyBundle {
...
-                let mut signer = Signer::new(MessageDigest::sha256(), &pkey)?;
-                signer.update(data)?;
-                return Ok(signer.sign_to_vec()?);
-            }
-            _ => {
-                return Err(RvError::ErrPkiKeyOperationInvalid);
-            }
+        let digest = match self.key_type.as_str() {
+            "rsa" | "ec" => MessageDigest::sha256(),
+            "sm2" => MessageDigest::sm3(),
```
Comment (InfoHunter) [inline src/utils/key.rs:136]: Same as above, no `#[cfg(not(tongsuo))]` needed?

File: src/utils/key.rs (hunk 1 above)
Comment (wa5i) [reply src/utils/key.rs:94]: Yes, fixed.

File: src/utils/key.rs (hunk 2 above)
Comment (wa5i) [reply src/utils/key.rs:138]: > Same as above, no `#[cfg(not(tongsuo))]` needed?

fixed.

**Review by InfoHunter — APPROVED** [review]

**Pull request merged**